src/main/java/com/relayrides/pushy/apns/ApnsClient.java

```diff
@@ -0,0 +1,749 @@
+     */
+    public Future<Void> getReconnectionFuture() {
+        final Future<Void> reconnectionFuture;
+
+        synchronized (this.bootstrap) {
+            if (this.isConnected()) {
+                reconnectionFuture = this.connectionReadyPromise.channel().newSucceededFuture();
+            } else if (this.reconnectionPromise != null) {
+                // If we're not connected, but have a reconnection promise, we're in the middle of a reconnection
+                // attempt.
+                reconnectionFuture = this.reconnectionPromise;
+            } else {
+                // We're not connected and have no reconnection future, which means we've either never connected or have
+                // explicitly disconnected.
+                reconnectionFuture = new FailedFuture<Void>(GlobalEventExecutor.INSTANCE,
+                        new IllegalStateException("Client was not previously connected."));
+            }
+        }
+
+        return reconnectionFuture;
+    }
+
+    /**
+     * <p>Sends a push notification to the APNs gateway.</p>
+     *
+     * <p>This method returns a {@code Future} that indicates whether the notification was accepted or rejected by the
+     * gateway. If the notification was accepted, it may be delivered to its destination device at some time in the
+     * future, but final delivery is not guaranteed. Rejections should be considered permanent failures, and callers
+     * should <em>not</em> attempt to re-send the notification.</p>
+     *
+     * <p>The returned {@code Future} may fail with an exception if the notification could not be sent. Failures to
+     * <em>send</em> a notification to the gateway—i.e. those that fail with exceptions—should generally be considered
+     * non-permanent, and callers should attempt to re-send the notification when the underlying problem has been
+     * resolved.</p>
+     *
+     * <p>In particular, attempts to send a notification when the client is not connected will fail with a
+     * {@link ClientNotConnectedException}. If the client was previously connected and has not been explicitly
+     * disconnected (via the {@link ApnsClient#disconnect()} method), the client will attempt to reconnect
+     * automatically. Callers may wait for a reconnection attempt to complete by waiting for the {@code Future} returned
+     * by the {@link ApnsClient#getReconnectionFuture()} method.</p>
+     *
+     * @param notification the notification to send to the APNs gateway
+     *
+     * @return a {@code Future} that will complete when the notification has been either accepted or rejected by the
+     * APNs gateway
+     */
+    public Future<PushNotificationResponse<T>> sendNotification(final T notification) {
+
+        final Future<PushNotificationResponse<T>> responseFuture;
+
+        // Instead of synchronizing here, we keep a final reference to the connection ready promise. We can get away
```

Comment:
It could, but we still run the risk of a `volatile` promise getting set to `null` by another thread halfway through this method (i.e. the first read will be non-null, but subsequent reads will be null and cause exceptions).

The local reference approach means that the state of the promise might change (i.e. the connection might close), but at least it won't mysteriously go `null` on us.